### PR TITLE
Fix not resolving type params in object and record types

### LINKED ISSUE #35596: Resolve type params in object and record types in `TypeParamResolver`
Currently the type params within object and record types aren't resolved in https://github.com/ballerina-platform/ballerina-lang/blob/95f08123f94dafc0e1a7758917f6f6822b6c83c1/compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/TypeParamResolver.java#L153. Need to implement this to address issues like the following:
Screenshot 2022-03-28 at 14 41 37
Note how the type params within the object type is not resolved.

### PR DESCRIPTION
**Pull request by dulajdilshan**
## Purpose
$title

Fixes #35596 

## Remarks
Need to address the following as well
- https://github.com/ballerina-platform/ballerina-lang/issues/36519
- https://github.com/ballerina-platform/ballerina-lang/issues/36517

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/TypeParamResolver.java [hunk 1]
```diff
@@ -151,12 +160,60 @@ public BType visit(BArrayType typeInSymbol, BType boundType) {
...
+            BAttachedFunction newAttachedFunc = new BAttachedFunction(attachedFunc.funcName, newInvokableSymbol,
+                                                    (BInvokableType) newType, attachedFunc.pos);
+            newAttachedFuncs.add(newAttachedFunc);
+        }
+
+        BObjectTypeSymbol newTypeSymbol = new BObjectTypeSymbol(objectTypeSymbol.tag, objectTypeSymbol.flags,
+                objectTypeSymbol.name, objectTypeSymbol.pkgID, objectTypeSymbol.getType(), objectTypeSymbol.owner,
+                objectTypeSymbol.pos, objectTypeSymbol.origin);
+        BObjectType newObjectType = new BObjectType(newTypeSymbol, typeInSymbol.flags);
+
```
Comment (sanjana) [inline compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/TypeParamResolver.java:190]:
```suggestion
```
Remove the superfluous new line

File: compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/TypeParamResolver.java [hunk 2]
```diff
@@ -151,12 +160,60 @@ public BType visit(BArrayType typeInSymbol, BType boundType) {
...
+        return newObjectType;
     }
 
     @Override
     public BType visit(BRecordType typeInSymbol, BType boundType) {
-        return typeInSymbol;
+
+        LinkedHashMap<String, BField> newRecordFields = new LinkedHashMap<>();
+
```
Comment (sanjana) [inline compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/TypeParamResolver.java:201]:
```suggestion
```

Ditto

File: compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/TypeParamResolver.java [hunk 3]
```diff
@@ -151,12 +160,60 @@ public BType visit(BArrayType typeInSymbol, BType boundType) {
...
+        for (Map.Entry<String, BField> entry: entries) {
+            BType newType = resolve(entry.getValue().type, boundType);
+            BVarSymbol newVarSymbol = createNewVarSymbol(entry.getValue().symbol, newType);
+            BField newField = new BField(newVarSymbol.getName(), newVarSymbol.getPosition(), newVarSymbol);
+            newRecordFields.put(entry.getKey(), newField);
+        }
+
+        BType newRestType = resolve(typeInSymbol.restFieldType, boundType);
+        BRecordType newRecordType = new BRecordType(typeInSymbol.tsymbol, typeInSymbol.flags);
+
```
Comment (sanjana) [inline compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/TypeParamResolver.java:212]:
```suggestion
```

Ditto

File: compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/TypeParamResolver.java [hunk 4]
```diff
@@ -267,4 +324,67 @@ public BType visit(BTypeReferenceType typeInSymbol, BType s) {
...
+        for (BVarSymbol param : symbol.params) {
+            BType newParamType = resolve(param.type, boundType);
+            BVarSymbol newVarSymbol = createNewVarSymbol(param, newParamType);
+            newInvokableSymbol.params.add(newVarSymbol);
+        }
+        if (symbol.restParam != null) {
+            BType newReturnType = resolve(symbol.restParam.type, boundType);
+            newInvokableSymbol.restParam = createNewVarSymbol(symbol.restParam, newReturnType);
+        }
+        newInvokableSymbol.retType = resolve(symbol.retType, boundType);
```
Comment (sanjana) [inline compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/TypeParamResolver.java:359]:
```suggestion
        }
        
        newInvokableSymbol.retType = resolve(symbol.retType, boundType);
```

File: compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/TypeParamResolver.java [hunk 5]
```diff
@@ -267,4 +324,67 @@ public BType visit(BTypeReferenceType typeInSymbol, BType s) {
...
+    private BInvokableSymbol resolveInvokableSymbol(BInvokableSymbol symbol, BType newInvokableType, BType boundType) {
+        BInvokableSymbol newInvokableSymbol = duplicateSymbol(symbol);
+        newInvokableSymbol.type = newInvokableType;
+
+        for (BVarSymbol param : symbol.params) {
+            BType newParamType = resolve(param.type, boundType);
+            BVarSymbol newVarSymbol = createNewVarSymbol(param, newParamType);
+            newInvokableSymbol.params.add(newVarSymbol);
+        }
+        if (symbol.restParam != null) {
```
Comment (sanjana) [inline compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/TypeParamResolver.java:355]:
```suggestion
        }
        
        if (symbol.restParam != null) {
```
Shouldn't there be a new line here?

**Review by sanjana — DISMISSED** [review]
LS test cases should be updated as well.

**Review by DamithaSenevirathne — APPROVED** [review]
LGTM

**Pull request merged**